### PR TITLE
Feature: Add Ferdium Translator

### LINKED ISSUE #504: Feature: add "translate" option for selected text
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/ferdium/ferdium-app/blob/develop/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/ferdium/ferdium-app/blob/develop/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/ferdium/ferdium-app/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Add option to translate selected text in any service.

### Proposed Solution

It would be a very nice feature to be able to select some text on a page on any service and then right-click and be able to see an automatic translation in a popup. Pretty much like the Google Translate Chrome extension works.

Optionally, if the text exceeds a certain length, a link to Google Translate could be provided and the selected text passed as a parameter to that page.

image

### Alternatives Considered

An alternative solution is to install the Google Translate service and to copy-paste the text. This works but it requires much more effort.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by SpecialAro**
#### Pre-flight Checklist

1. Please remember that if you are logging a bug for some service that has _stopped working_ or is _working incorrectly_, please log the bug [here](https://github.com/ferdium/ferdium-recipes/issues)
2. If you are requesting support for a **new service** in Ferdium, please log it [here](https://github.com/ferdium/ferdium-recipes/pulls)
3. Please remember to read the [self-help documentation](https://github.com/ferdium/ferdium-app#troubleshooting-recipes-self-help) - in case it helps you unblock yourself for issues related to older versions of recipes that were installed on your machine. (These will get automatically upgraded when you upgrade to the newer versions of Ferdium, but to get new recipes between Ferdium releases, this documentation is quite useful.)
4. Please ensure you've completed all of the following.

- [x] I have read the [Contributing Guidelines](https://github.com/ferdium/ferdium-app/blob/develop/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/ferdium/ferdium-app/blob/develop/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://github.com/ferdium/ferdium-app/issues) for a feature request that matches the one I want to file, without success.

#### Description of Change
Add feature to translate text natively using https://github.com/shikar/NODE_GOOGLE_TRANSLATE package and a LibreTranslate self-hosted option (already running on our server on https://translator.ferdium.org - maybe we can disable the web ui).

#### Motivation and Context
This fixes https://github.com/ferdium/ferdium-app/issues/504.
With further implementation we can maybe provide a field for the user to input the URL of a self-hosted instance of LibreTranslate they have.

#### Screenshots
Ferdium

#### Checklist

- [x] My pull request is properly named
- [x] The changes respect the code style of the project (`npm run prepare-code`)
- [x] `npm test` passes
- [x] I tested/previewed my changes locally

#### Release Notes

notes: Added native Ferdium Translator.

### CODE REVIEW

**Comment by kris7t** [comment]
We should make it very clear to users _when_ and _how_ does this leak information (i.e., the text they selected) to google.

Ideally, we'd use a local-only solution like https://www.mozilla.org/en-US/firefox/features/translate/, but afaik this is not (yet) possible within electron. :disappointed:

File: src/containers/settings/EditSettingsScreen.tsx
Comment (vraravam) [inline src/containers/settings/EditSettingsScreen.tsx]: Please change this to "Default Translator language"

File: src/i18n/locales/en-US.json
Comment (vraravam) [inline src/i18n/locales/en-US.json]: please change this to "Default Translator language"

**Comment by SpecialAro** [comment]
> We should make it very clear to users _when_ and _how_ does this leak information (i.e., the text they selected) to google.

Totally agree with you!
As per implementation, it only leaks to google (I think anonymously - since I don't parse any more information than the string to translate and the language to translate - tho, they probably can try to track the IP or so...) only when the user clicks the button after selecting the text. Where should we have this information?

**Review by vraravam — APPROVED** [review]
LGTM

**Comment by SpecialAro** [comment]
> LGTM

@vraravam (how) does this PR interfere with the native macOS translator (or is it just the dictionary, can't remember right now)? Should we disable this context menu to macOS users?

**Comment by kris7t** [comment]
> Where should we have this information?

One way to do this would be an explicit opt-in switch in _Settings > Privacy_ (with a note that translations are provided by google translate). Maybe we could eventually add a backend selector, too, to enable the use proxies like https://lingva.ml/

Otoh, making this explicitly opt-in might have discoverability issues.

**Comment by vraravam** [comment]
> @vraravam (how) does this PR interfere with the native macOS translator (or is it just the dictionary, can't remember right now)? Should we disable this context menu to macOS users?

That's only for the spell checker on macos - not a translator afaik.

**Comment by SpecialAro** [comment]
> > Where should we have this information?
> 
> One way to do this would be an explicit opt-in switch in _Settings > Privacy_ (with a note that translations are provided by google translate). Maybe we could eventually add a backend selector, too, to enable the use proxies like https://lingva.ml/
> 
> Otoh, making this explicitly opt-in might have discoverability issues.

I've been thinking about this and google translator is, in fact, kinda "concerning" when you look at their privacy policy on this. Maybe we can implement a provider drop-down too for this, which could include Google Translate and LibreTranslate (as default) - which we could self host on our servers. Do you think this option is viable?

And also, we should have a toggle to disable this feature entirely.

Therefore, maybe don't merge this just yet (if you agree with me), I'll try to make some changes tomorrow and see what works best 😁

**Comment by SpecialAro** [comment]
Hey!

Ok, had some problems with the implementation because of the use of "fetch" inside the webview (which caused a bunch of problems with the Content Security Policy of each website).
Therefore, I chose a different approach using IPC to expose things to the main process. In there, I make all the needed things and return an object with string (translated text) and boolean status (if it translated or catch an error), that are used back in the render process.

I think this solution is "cleaner", but let me know what you think. Also, there is missing the addition of tests for `translation-helpers.ts`, if anyone can contribute towards that I would be really grateful (I'm really tired, need to rest a bit).

Also, this feature is disabled by default... but let me know if you think we should change that. BTW, Ferdium Translator is the default Translator as well.

Hope everything is ok! Please review this PR again before merging.

Thank you 😄

File: src/config.ts
Comment (vraravam) [inline src/config.ts]: typo in the constant name?

**Review by vraravam — CHANGES_REQUESTED** [review]

**Review by vraravam — APPROVED** [review]
LGTM!

File: src/webview/contextMenuBuilder.ts
Comment (cino) [inline src/webview/contextMenuBuilder.ts]: Can we extend the ContextMenuParams type and add our own properties? So we don't need to use these ts-expect-errors

File: src/webview/contextMenuBuilder.ts
Comment (cino) [inline src/webview/contextMenuBuilder.ts]: Same

File: src/webview/contextMenuBuilder.ts
Comment (cino) [inline src/webview/contextMenuBuilder.ts]: Same

File: src/webview/contextMenuBuilder.ts
Comment (SpecialAro) [reply src/webview/contextMenuBuilder.ts]: I agree... But I've never done so. Can you give me an example?

I'm trying to do this:
```
interface IProps extends Electron.ContextMenuParams {
  info: any;
  menuInfo: {
    linkURL: any;
    linkText: string;
    clipboardNotifications: any;
    dictionarySuggestions: any;
    misspelledWord: string;
    selectionText: string;
    searchEngine: string;
    translatorLanguage: string;
    srcURL: string;
    editFlags: any;
    hasImageContents: boolean;
    x: number;
    y: number;
    pageURL: any;
    enableTranslator: boolean;
    translatorEngine: string;
  };
}

export default class FerdiumContextMenuParams implements IProps {
  x: number;

  y: number;

  frame: Electron.WebFrameMain;

  linkURL: string;

  linkText: string;

  pageURL: string;

  frameURL: string;

  srcURL: string;

  mediaType:
    | 'none'
    | 'image'
    | 'audio'
    | 'video'
    | 'canvas'
    | 'file'
    | 'plugin';

  hasImageContents: boolean;

  isEditable: boolean;

  selectionText: string;

  titleText: string;

  altText: string;

  suggestedFilename: string;

  selectionRect: Electron.Rectangle;

  selectionStartOffset: number;

  referrerPolicy: Electron.Referrer;

  misspelledWord: string;

  dictionarySuggestions: string[];

  frameCharset: string;

  inputFieldType: string;

  spellcheckEnabled: boolean;

  menuSourceType:
    | 'none'
    | 'mouse'
    | 'keyboard'
    | 'touch'
    | 'touchMenu'
    | 'longPress'
    | 'longTap'
    | 'touchHandle'
    | 'stylus'
    | 'adjustSelection'
    | 'adjustSelectionReset';

  mediaFlags: Electron.MediaFlags;

  editFlags: Electron.EditFlags;

  info: any;

  menuInfo: any;
}
```

But it is throwing me a bunch of errors:
image

**Comment by SpecialAro** [comment]
@cino just committed some changes regarding your previous comments. I'm not sure if this is the proper way to do this so I'll wait for your approval before merging this PR, if that's ok with you 😄 

Thank you so much in advance!

File: src/helpers/translation-helpers.ts
Comment (cino) [inline src/helpers/translation-helpers.ts]: Can this be removed?

File: src/helpers/translation-helpers.ts
Comment (cino) [inline src/helpers/translation-helpers.ts]: return type

**Review by cino — APPROVED** [review]

**Comment by gogosese** [comment]
Have you ever encountered difficulty in translating during a conversation with your friends?

**Pull request merged**
